File: templates/rust/protocols/command_protocol.rs
//! MAVLink command protocol client and server.

use std::sync::Arc;
use std::time::Duration;

use futures::StreamExt;
use tokio::task::JoinHandle;

use crate::{CommandAck, CommandInt, CommandLong, MavCmd, MavResult};

use super::mavlink_cancellation::MavlinkCancellationToken;
use super::mavlink_session::{MavlinkSession, SessionWaitError};

/// GCS-side MAVLink command protocol client.
pub struct CommandProtocol {
    session: Arc<MavlinkSession>,
    target_system: u8,
    target_component: u8,
    default_timeout: Duration,
}

impl CommandProtocol {
    pub fn new(
        session: Arc<MavlinkSession>,
        target_system: u8,
        target_component: u8,
        default_timeout: Duration,
    ) -> Self {
        Self {
            session,
            target_system,
            target_component,
            default_timeout,
        }
    }

    pub async fn send_long(
        &self,
        command: CommandLong,
        timeout: Option<Duration>,
        cancel: Option<&MavlinkCancellationToken>,
    ) -> Result<CommandAck, SessionWaitError> {
        let cmd = command.command;
        self.session.send(Box::new(command)).await?;
        self.wait_for_ack(cmd, timeout, cancel).await
    }

    pub async fn send_int(
        &self,
        command: CommandInt,
        timeout: Option<Duration>,
        cancel: Option<&MavlinkCancellationToken>,
    ) -> Result<CommandAck, SessionWaitError> {
        let cmd = command.command;
        self.session.send(Box::new(command)).await?;
        self.wait_for_ack(cmd, timeout, cancel).await
    }

    pub async fn command_long(
        &self,
        command: MavCmd,
        param1: f32,
        param2: f32,
        param3: f32,
        param4: f32,
        param5: f32,
        param6: f32,
        param7: f32,
        confirmation: u8,
        timeout: Option<Duration>,
        cancel: Option<&MavlinkCancellationToken>,
    ) -> Result<CommandAck, SessionWaitError> {
        self.send_long(
            CommandLong {
                param1,
                param2,
                param3,
                param4,
                param5,
                param6,
                param7,
                command,
                target_system: self.target_system,
                target_component: self.target_component,
                confirmation,
            },
            timeout,
            cancel,
        )
        .await
    }

    pub async fn request_message(
        &self,
        message_id: u32,
        param2: f32,
        timeout: Option<Duration>,
        cancel: Option<&MavlinkCancellationToken>,
    ) -> Result<CommandAck, SessionWaitError> {
        self.command_long(
            MavCmd::MAV_CMD_REQUEST_MESSAGE,
            message_id as f32,
            param2,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0,
            timeout,
            cancel,
        )
        .await
    }

    pub async fn set_message_interval(
        &self,
        message_id: u32,
        interval_us: u32,
        timeout: Option<Duration>,
        cancel: Option<&MavlinkCancellationToken>,
    ) -> Result<CommandAck, SessionWaitError> {
        self.command_long(
            MavCmd::MAV_CMD_SET_MESSAGE_INTERVAL,
            message_id as f32,
            interval_us as f32,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0,
            timeout,
            cancel,
        )
        .await
    }

    pub async fn stop_message_interval(
        &self,
        message_id: u32,
        timeout: Option<Duration>,
        cancel: Option<&MavlinkCancellationToken>,
    ) -> Result<CommandAck, SessionWaitError> {
        self.set_message_interval(message_id, 0, timeout, cancel)
            .await
    }

    pub async fn set_mission_current(
        &self,
        sequence: u16,
        reset_mission: bool,
        timeout: Option<Duration>,
        cancel: Option<&MavlinkCancellationToken>,
    ) -> Result<CommandAck, SessionWaitError> {
        self.command_long(
            MavCmd::MAV_CMD_DO_SET_MISSION_CURRENT,
            f32::from(sequence),
            if reset_mission { 1.0 } else { 0.0 },
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0,
            timeout,
            cancel,
        )
        .await
    }

    pub async fn arm(
        &self,
        force: bool,
        timeout: Option<Duration>,
        cancel: Option<&MavlinkCancellationToken>,
    ) -> Result<CommandAck, SessionWaitError> {
        self.command_long(
            MavCmd::MAV_CMD_COMPONENT_ARM_DISARM,
            1.0,
            if force { 21196.0 } else { 0.0 },
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0,
            timeout,
            cancel,
        )
        .await
    }

    pub async fn disarm(
        &self,
        force: bool,
        timeout: Option<Duration>,
        cancel: Option<&MavlinkCancellationToken>,
    ) -> Result<CommandAck, SessionWaitError> {
        self.command_long(
            MavCmd::MAV_CMD_COMPONENT_ARM_DISARM,
            0.0,
            if force { 21196.0 } else { 0.0 },
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0,
            timeout,
            cancel,
        )
        .await
    }

    pub async fn takeoff(
        &self,
        altitude: f32,
        timeout: Option<Duration>,
        cancel: Option<&MavlinkCancellationToken>,
    ) -> Result<CommandAck, SessionWaitError> {
        self.command_long(
            MavCmd::MAV_CMD_NAV_TAKEOFF,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            altitude,
            0,
            timeout,
            cancel,
        )
        .await
    }

    pub async fn land(
        &self,
        timeout: Option<Duration>,
        cancel: Option<&MavlinkCancellationToken>,
    ) -> Result<CommandAck, SessionWaitError> {
        self.command_long(
            MavCmd::MAV_CMD_NAV_LAND,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0,
            timeout,
            cancel,
        )
        .await
    }

    pub async fn return_to_launch(
        &self,
        timeout: Option<Duration>,
        cancel: Option<&MavlinkCancellationToken>,
    ) -> Result<CommandAck, SessionWaitError> {
        self.command_long(
            MavCmd::MAV_CMD_NAV_RETURN_TO_LAUNCH,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0,
            timeout,
            cancel,
        )
        .await
    }

    pub async fn wait_for_ack(
        &self,
        command: MavCmd,
        timeout: Option<Duration>,
        cancel: Option<&MavlinkCancellationToken>,
    ) -> Result<CommandAck, SessionWaitError> {
        self.session
            .wait_for_message(
                move |message| {
                    if let Some(ack) = (message as &dyn std::any::Any).downcast_ref::<CommandAck>() {
                        return ack.command == command;
                    }
                    false
                },
                Some(self.target_system),
                None,
                timeout.unwrap_or(self.default_timeout),
                cancel,
            )
            .await
            .and_then(|message| {
                (message.as_ref() as &dyn std::any::Any)
                    .downcast_ref::<CommandAck>()
                    .cloned()
                    .ok_or(SessionWaitError::Closed)
            })
    }
}

pub type CommandLongHandler =
    dyn Fn(CommandLong) -> std::pin::Pin<Box<dyn std::future::Future<Output = MavResult> + Send>>
    + Send
    + Sync;
pub type CommandIntHandler =
    dyn Fn(CommandInt) -> std::pin::Pin<Box<dyn std::future::Future<Output = MavResult> + Send>>
    + Send
    + Sync;

/// Vehicle-side command handler.
pub struct CommandServer {
    session: Arc<MavlinkSession>,
    on_command_long: Option<Arc<CommandLongHandler>>,
    on_command_int: Option<Arc<CommandIntHandler>>,
    reader: JoinHandle<()>,
}

impl CommandServer {
    pub fn new(
        session: Arc<MavlinkSession>,
        on_command_long: Option<Arc<CommandLongHandler>>,
        on_command_int: Option<Arc<CommandIntHandler>>,
    ) -> Self {
        let session_reader = Arc::clone(&session);
        let long_handler = on_command_long.clone();
        let int_handler = on_command_int.clone();
        let reader = tokio::spawn(async move {
            let mut frames = session_reader.frames();
            while let Some(frame) = frames.next().await {
                let _ = handle_command_frame(&session_reader, long_handler.as_deref(), int_handler.as_deref(), frame)
                    .await;
            }
        });
        Self {
            session,
            on_command_long,
            on_command_int,
            reader,
        }
    }

    pub async fn close(self) {
        self.reader.abort();
    }
}

async fn handle_command_frame(
    session: &MavlinkSession,
    on_command_long: Option<&CommandLongHandler>,
    on_command_int: Option<&CommandIntHandler>,
    frame: Arc<crate::mavlink_frame::MavlinkFrame>,
) -> Result<(), SessionWaitError> {
    if let Some(command) = (frame.message.as_ref() as &dyn std::any::Any).downcast_ref::<CommandLong>() {
        if command.target_system != session.system_id() {
            return Ok(());
        }
        let result = if let Some(handler) = on_command_long {
            handler(command.clone()).await
        } else {
            MavResult::MAV_RESULT_ACCEPTED
        };
        send_ack(session, &frame, command.command, result).await?;
        return Ok(());
    }

    if let Some(command) = (frame.message.as_ref() as &dyn std::any::Any).downcast_ref::<CommandInt>() {
        if command.target_system != session.system_id() {
            return Ok(());
        }
        let result = if let Some(handler) = on_command_int {
            handler(command.clone()).await
        } else {
            MavResult::MAV_RESULT_ACCEPTED
        };
        send_ack(session, &frame, command.command, result).await?;
    }

    Ok(())
}

async fn send_ack(
    session: &MavlinkSession,
    frame: &Arc<crate::mavlink_frame::MavlinkFrame>,
    command: MavCmd,
    result: MavResult,
) -> Result<(), SessionWaitError> {
    session
        .send(Box::new(CommandAck {
            command,
            result,
            progress: 0,
            result_param2: 0,
            target_system: frame.system_id,
            target_component: frame.component_id,
        }))
        .await
}

File: templates/py/protocols/command_protocol.py
from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable
from datetime import timedelta

from mavlink import CommandAck, CommandInt, CommandLong, MavCmd, MavResult
from mavlink_frame import MavlinkFrame

from .mavlink_cancellation import MavlinkCancellationToken
from .mavlink_session import MavlinkSession


class CommandProtocol:
    def __init__(
        self,
        session: MavlinkSession,
        target_system: int,
        target_component: int,
        default_timeout: timedelta = timedelta(seconds=5),
    ) -> None:
        self.session = session
        self.target_system = target_system
        self.target_component = target_component
        self.default_timeout = default_timeout

    async def send_long(
        self,
        command: CommandLong,
        timeout: timedelta | None = None,
        cancel: MavlinkCancellationToken | None = None,
    ) -> CommandAck:
        await self.session.send(command)
        return await self.wait_for_ack(command.command, timeout=timeout, cancel=cancel)

    async def send_int(
        self,
        command: CommandInt,
        timeout: timedelta | None = None,
        cancel: MavlinkCancellationToken | None = None,
    ) -> CommandAck:
        await self.session.send(command)
        return await self.wait_for_ack(command.command, timeout=timeout, cancel=cancel)

    async def command_long(
        self,
        command: MavCmd,
        param1: float = 0,
        param2: float = 0,
        param3: float = 0,
        param4: float = 0,
        param5: float = 0,
        param6: float = 0,
        param7: float = 0,
        confirmation: int = 0,
        timeout: timedelta | None = None,
        cancel: MavlinkCancellationToken | None = None,
    ) -> CommandAck:
        return await self.send_long(
            CommandLong(
                param1=param1,
                param2=param2,
                param3=param3,
                param4=param4,
                param5=param5,
                param6=param6,
                param7=param7,
                command=command,
                target_system=self.target_system,
                target_component=self.target_component,
                confirmation=confirmation,
            ),
            timeout=timeout,
            cancel=cancel,
        )

    async def request_message(
        self,
        message_id: int,
        param2: float = 0,
        timeout: timedelta | None = None,
        cancel: MavlinkCancellationToken | None = None,
    ) -> CommandAck:
        return await self.command_long(
            command=MavCmd.MAV_CMD_REQUEST_MESSAGE,
            param1=float(message_id),
            param2=param2,
            timeout=timeout,
            cancel=cancel,
        )

    async def set_message_interval(
        self,
        message_id: int,
        interval_us: int,
        timeout: timedelta | None = None,
        cancel: MavlinkCancellationToken | None = None,
    ) -> CommandAck:
        return await self.command_long(
            command=MavCmd.MAV_CMD_SET_MESSAGE_INTERVAL,
            param1=float(message_id),
            param2=float(interval_us),
            timeout=timeout,
            cancel=cancel,
        )

    async def stop_message_interval(
        self,
        message_id: int,
        timeout: timedelta | None = None,
        cancel: MavlinkCancellationToken | None = None,
    ) -> CommandAck:
        return await self.set_message_interval(
            message_id, 0, timeout=timeout, cancel=cancel
        )

    async def set_mission_current(
        self,
        sequence: int,
        reset_mission: bool = False,
        timeout: timedelta | None = None,
        cancel: MavlinkCancellationToken | None = None,
    ) -> CommandAck:
        return await self.command_long(
            command=MavCmd.MAV_CMD_DO_SET_MISSION_CURRENT,
            param1=float(sequence),
            param2=1.0 if reset_mission else 0.0,
            timeout=timeout,
            cancel=cancel,
        )

    async def arm(
        self,
        force: bool = False,
        timeout: timedelta | None = None,
        cancel: MavlinkCancellationToken | None = None,
    ) -> CommandAck:
        return await self.command_long(
            command=MavCmd.MAV_CMD_COMPONENT_ARM_DISARM,
            param1=1,
            param2=21196 if force else 0,
            timeout=timeout,
            cancel=cancel,
        )

    async def disarm(
        self,
        force: bool = False,
        timeout: timedelta | None = None,
        cancel: MavlinkCancellationToken | None = None,
    ) -> CommandAck:
        return await self.command_long(
            command=MavCmd.MAV_CMD_COMPONENT_ARM_DISARM,
            param1=0,
            param2=21196 if force else 0,
            timeout=timeout,
            cancel=cancel,
        )

    async def takeoff(
        self,
        altitude: float = 10,
        timeout: timedelta | None = None,
        cancel: MavlinkCancellationToken | None = None,
    ) -> CommandAck:
        return await self.command_long(
            command=MavCmd.MAV_CMD_NAV_TAKEOFF,
            param7=altitude,
            timeout=timeout,
            cancel=cancel,
        )

    async def land(
        self,
        timeout: timedelta | None = None,
        cancel: MavlinkCancellationToken | None = None,
    ) -> CommandAck:
        return await self.command_long(
            command=MavCmd.MAV_CMD_NAV_LAND,
            timeout=timeout,
            cancel=cancel,
        )

    async def return_to_launch(
        self,
        timeout: timedelta | None = None,
        cancel: MavlinkCancellationToken | None = None,
    ) -> CommandAck:
        return await self.command_long(
            command=MavCmd.MAV_CMD_NAV_RETURN_TO_LAUNCH,
            timeout=timeout,
            cancel=cancel,
        )

    async def wait_for_ack(
        self,
        command: MavCmd,
        timeout: timedelta | None = None,
        cancel: MavlinkCancellationToken | None = None,
    ) -> CommandAck:
        message = await self.session.wait_for_message(
            predicate=lambda msg: isinstance(msg, CommandAck) and msg.command == command,
            from_system_id=self.target_system,
            timeout=timeout or self.default_timeout,
            cancel=cancel,
        )
        assert isinstance(message, CommandAck)
        return message


CommandLongHandler = Callable[[CommandLong], Awaitable[MavResult] | MavResult]
CommandIntHandler = Callable[[CommandInt], Awaitable[MavResult] | MavResult]


class CommandServer:
    def __init__(
        self,
        session: MavlinkSession,
        on_command_long: CommandLongHandler | None = None,
        on_command_int: CommandIntHandler | None = None,
    ) -> None:
        self.session = session
        self.on_command_long = on_command_long
        self.on_command_int = on_command_int
        self._frame_task = asyncio.create_task(self._listen_frames())

    async def close(self) -> None:
        self._frame_task.cancel()
        try:
            await self._frame_task
        except asyncio.CancelledError:
            pass

    async def _listen_frames(self) -> None:
        async for frame in self.session.frames():
            await self._on_frame(frame)

    async def _on_frame(self, frame: MavlinkFrame) -> None:
        message = frame.message

        if isinstance(message, CommandLong):
            if message.target_system != self.session.system_id:
                return
            if self.on_command_long is not None:
                result = self.on_command_long(message)
            else:
                result = MavResult.MAV_RESULT_ACCEPTED
            if asyncio.iscoroutine(result):
                result = await result
            await self._send_ack(frame, message.command, result)
            return

        if isinstance(message, CommandInt):
            if message.target_system != self.session.system_id:
                return
            if self.on_command_int is not None:
                result = self.on_command_int(message)
            else:
                result = MavResult.MAV_RESULT_ACCEPTED
            if asyncio.iscoroutine(result):
                result = await result
            await self._send_ack(frame, message.command, result)

    async def _send_ack(
        self, request_frame: MavlinkFrame, command: MavCmd, result: MavResult
    ) -> None:
        await self.session.send(
            CommandAck(
                command=command,
                result=result,
                progress=0,
                result_param2=0,
                target_system=request_frame.system_id,
                target_component=request_frame.component_id,
            )
        )
